### ISSUE TITLE
assert in xelb:unmarshal

### ISSUE THREAD

**Issue by razcampagne** [ch11ng/exwm#158]
Hi,

When trying and start some games from steam, xelb sometimes trigger the following assert in xelb:unmarshal:

(cl-assert (= (* 4 value-len) (length tmp)))

The issue seems to occur when (= (length value) bytes-after). This would mean that the statement
(setq tmp (substring value 0 (- (length value) bytes-after))) will set an empty vector to tmp.

As a workaround, I added (when (> (length tmp) 0) ...) as a test surrounding (pcase format ...) and that did the trick.

As I have to idea how xcb works, I wouldn't dare submitting a patch, but if you think this is a reasonable fix, please do use it.

If you need more information about the faulty window, just tell me how to get such information.

Thanks!

**Comment by wacossusca34** [comment]
can confirm that I have the (exact) same issue

**Comment by ch11ng** [comment]
This issue was reported before (ch11ng/xelb#10).  I've just reread the X protocol and apparently I misunderstood the meaning of the `bytes-after` field in the `GetProperty` request.  Though the fix can be easy, I still want to know which request triggers the bug; there might be another bug where a property is not read entirely.  Below are possible candidates:

```elisp
xcb:icccm:-GetProperty~reply
  +--xcb:ewmh:get-_NET_WM_OPAQUE_REGION~reply
  +--xcb:ewmh:-get-_NET_WM_ICON~reply
  +--xcb:ewmh:get-_NET_WM_ALLOWED_ACTIONS~reply
  +--xcb:ewmh:get-_NET_WM_STATE~reply
  +--xcb:ewmh:get-_NET_WM_WINDOW_TYPE~reply
  +--xcb:ewmh:get-_NET_VIRTUAL_ROOTS~reply
  +--xcb:ewmh:get-_NET_WORKAREA~reply
  +--xcb:ewmh:get-_NET_DESKTOP_VIEWPORT~reply
  +--xcb:ewmh:get-_NET_CLIENT_LIST_STACKING~reply
  +--xcb:ewmh:get-_NET_CLIENT_LIST~reply
  +--xcb:ewmh:get-_NET_SUPPORTED~reply
  +--xcb:icccm:get-WM_COLORMAP_WINDOWS~reply
  +--xcb:icccm:get-WM_PROTOCOLS~reply
  +--xcb:icccm:-GetProperty-explicit~reply
  |    +--xcb:ewmh:get-_NET_WM_FULLSCREEN_MONITORS~reply
  |    +--xcb:ewmh:get-_NET_WM_SYNC_REQUEST_COUNTER~reply
  |    +--xcb:ewmh:get-_NET_FRAME_EXTENTS~reply
  |    +--xcb:ewmh:get-_NET_WM_ICON_GEOMETRY~reply
  |    +--xcb:ewmh:get-_NET_WM_STRUT_PARTIAL~reply
  |    +--xcb:ewmh:get-_NET_WM_STRUT~reply
  |    +--xcb:ewmh:get-_NET_DESKTOP_LAYOUT~reply
  |    +--xcb:ewmh:get-_NET_DESKTOP_GEOMETRY~reply
  |    +--xcb:icccm:get-WM_ICON_SIZE~reply
  |    +--xcb:icccm:get-WM_STATE~reply
  |    +--xcb:icccm:get-WM_HINTS~reply
  |    +--xcb:icccm:get-WM_SIZE_HINTS~reply
  |         +--xcb:icccm:get-WM_NORMAL_HINTS~reply
  +--xcb:icccm:-GetProperty-single~reply
  |    +--xcb:ewmh:get-_NET_WM_BYPASS_COMPOSITOR~reply
  |    +--xcb:ewmh:get-_NET_WM_USER_TIME_WINDOW~reply
  |    +--xcb:ewmh:get-_NET_WM_USER_TIME~reply
  |    +--xcb:ewmh:get-_NET_WM_HANDLED_ICONS~reply
  |    +--xcb:ewmh:get-_NET_WM_PID~reply
  |    +--xcb:ewmh:get-_NET_WM_DESKTOP~reply
  |    +--xcb:ewmh:get-_NET_SHOWING_DESKTOP~reply
  |    +--xcb:ewmh:get-_NET_SUPPORTING_WM_CHECK~reply
  |    +--xcb:ewmh:get-_NET_ACTIVE_WINDOW~reply
  |    +--xcb:ewmh:get-_NET_CURRENT_DESKTOP~reply
  |    +--xcb:ewmh:get-_NET_NUMBER_OF_DESKTOPS~reply
  |    +--xcb:icccm:get-WM_TRANSIENT_FOR~reply
  +--xcb:icccm:-GetProperty-text~reply
       +--xcb:ewmh:-GetProperty-utf8~reply
       |    +--xcb:ewmh:get-_NET_WM_VISIBLE_ICON_NAME~reply
       |    +--xcb:ewmh:get-_NET_WM_ICON_NAME~reply
       |    +--xcb:ewmh:get-_NET_WM_VISIBLE_NAME~reply
       |    +--xcb:ewmh:get-_NET_WM_NAME~reply
       |    +--xcb:ewmh:get-_NET_DESKTOP_NAMES~reply
       +--xcb:icccm:get-WM_CLIENT_MACHINE~reply
       +--xcb:icccm:get-WM_CLASS~reply
       +--xcb:icccm:get-WM_ICON_NAME~reply
       +--xcb:icccm:get-WM_NAME~reply
```

Could you insert `(setq debug-on-error t)` to your `.emacs` and then try to trigger this bug? It should give a `*Backtrace*` buffer of which the content can be helpful.

**Comment by razcampagne** [comment]
Oh sorry, I guess I reported to the wrong project.

Here is the backtrace you requested. The formatting is not great but hopefully, the information you need is somewhere in there.

https://gist.github.com/anonymous/066d0a9c5923cd94ccb252247665a358#file-exwm-unmarshal-backtrace-el

**Comment by ch11ng** [comment]
It turns out this issue is triggered by a wrongly set `_MOTIF_WM_HINTS` property.  I've checked the source of SDL2 and it seems it set the property correctly.  So I have no idea who set it.  Anyway I've fixed it in ch11ng/xelb@65b2934f69ffcb53906207b39d7624cb5cfc77de.
